### PR TITLE
feat(hybrid): Introduce RegionMapping method contracts

### PR DESCRIPTION
**Pull request by RyanSkonnord**
Set up a type system for region objects and the set of all regions, with placeholder methods for backend operations not implemented yet. These methods should serve as temporary interface contracts to unblock development on application code that would otherwise depend on those operations.

Specifically, the contracts cover:
* injecting the list of all regions from config (or whatever operational source of truth we settle on);
* looking up what region I'm in if I'm a region silo; and
* resolving an organization to its home region.

For the first practical use, I've dropped in a simple call to the second one, replacing some older placeholder code around "snowflake" ID generation.

### CODE REVIEW

File: src/sentry/types/region.py [hunk 1]
```diff
@@ -1,8 +1,120 @@
...
+    """
+
+    address: str
+    """The address of the region's silo.
+
+    In a production environment, this would typically be a subdomain (e.g.,
+    "eu.sentry.io"). In a development environment, it may be useful to define
+    region addresses as a custom hostname or port number (e.g,
+    "localhost:8001").
+    """
```
Comment (RyanSkonnord) [inline src/sentry/types/region.py:39]: The contract for the `address` attribute is a first pass and we'll probably need to iterate on it. In production, we'll definitely need to resolve regions into the subdomain where they're being hosted, and I wanted to generalize that to arbitrary hostnames and ports. But I have no idea whether it will make sense for the strings to look like `"eu.sentry.io"`, `"https://eu.sentry.io/"`, or whatever.

Note the assumption that the production config will be supplying a value of the form `"eu.sentry.io"` rather than just `"eu"`. If it's the latter, we'll additionally need to specify the `"sentry.io"` part somewhere -- I'd prefer to keep that domain-name-building logic outside of the application code as much as possible.

File: src/sentry/types/region.py [hunk 2]
```diff
@@ -1,8 +1,120 @@
...
+    category: RegionCategory
+    """The region's category."""
+
+    def __post_init__(self) -> None:
+        from sentry.utils.snowflake import REGION_ID
+
+        REGION_ID.validate(self.id)
+
+    def to_url(self, path: str) -> str:
+        return self.address + path
```
Comment (RyanSkonnord) [inline src/sentry/types/region.py:50]: As above, this is just conceptual. We'll handle "do we need to join them with a slash" and other annoying details when the time comes.

File: src/sentry/types/region.py [hunk 3]
```diff
@@ -1,8 +1,120 @@
...
+        local_region = self.get_by_name(local_region_name)
+        if not local_region:
+            raise Exception(f"SENTRY_REGION value not resolved to region: {local_region_name!r}")
+        return local_region
+
+
+_global_region_mapping = None
+
+
+def get_region_mapping() -> RegionMapping:
```
Comment (RyanSkonnord) [inline src/sentry/types/region.py:116]: This feels ugly. I'd prefer to do something like our common "registry" pattern where the region mapping is just initialized with the module, but that leads to circular imports because `sentry/conf.py` needs to import the `Region` and `RegionMapping` types.

Comment (asottile-sentry) [reply src/sentry/types/region.py:116]: a slightly less ugly version of this is to use `functools.lru_cache(maxsize=1)` to hide the `global` part

File: src/sentry/types/region.py (hunk 1 above)
Comment (mikejihbe) [reply src/sentry/types/region.py:39]: I think you're right that this might need some more thought. We will need to address non-API resources across regions as well, e.g. control silo reaching out to region silo kafka clusters. These might just be better as separate explicit configurations though

**Review by mikejihbe — COMMENTED** [review]
Directionally this is looking good 👍 , thanks for putting this together

File: src/sentry/types/region.py (hunk 3 above)
Comment (leeandher) [reply src/sentry/types/region.py:116]: I agree, actually getting the region with `get_region_mapping().get_local_region()` is a bit wordy. Maybe if this is a class and we shorten to `RegionMap`?
e.g.
```python
class RegionMap:
    def __new__(self):
        return (
            RegionMapping.load_from_config()
            if not _global_region_mapping
            else _global_region_mapping
        )
```
but it's essentially the same thing: `RegionMap().get_local_region()`

File: src/sentry/utils/snowflake.py [hunk 4]
```diff
@@ -72,10 +73,12 @@ def msb_0_ordering(value, width):
...
-    segment_values = {
-        VERSION_ID: msb_0_ordering(settings.SNOWFLAKE_VERSION_ID, VERSION_ID.length),
-        REGION_ID: settings.SNOWFLAKE_REGION_ID,
-    }
+    segment_values = {}
+
+    segment_values[VERSION_ID] = msb_0_ordering(settings.SNOWFLAKE_VERSION_ID, VERSION_ID.length)
+
+    local_region = get_region_mapping().get_local_region()
+    segment_values[REGION_ID] = local_region.id if local_region else 0
```
Comment (leeandher) [inline src/sentry/utils/snowflake.py:81]: Should we lock this `0` in a constant to denote a 'default region'? Just thinking that when snowflake IDs are rolled out into prod, we'd want 0 to map the the main US region, if we haven't set up the region config in SaaS yet.

Comment (leeandher) [reply src/sentry/utils/snowflake.py:81]: I guess we can think of it as 'default region' or 'unregioned' though, I think we should avoid SnowflakeIDs if we haven't set the region config just yet.

**Review by leeandher — APPROVED** [review]
👏 Looks good! I appreciate how documented and clear the API is. I cloned and spun this up locally too with no issues!

File: mypy.ini [hunk 5]
```diff
@@ -107,6 +107,7 @@ files = src/sentry/analytics/,
         src/sentry/tasks/symbolication.py,
```
Comment (corps) [inline mypy.ini:107]: Ah, interesting, so to be clear do all type-checked files need to be included here?

File: src/sentry/types/region.py (hunk 1 above)
Comment (corps) [reply src/sentry/types/region.py:39]: subdomain or redirect target for default HTTPS might be useful in general however.  This address may be like, "public endpoint"-eseque?

File: src/sentry/types/region.py (hunk 2 above)
Comment (corps) [reply src/sentry/types/region.py:50]: On the other hand, this definitely seems more related specifically to endpoint API and not necessary the Region as a whole.  I imagine there being a "SiloHTTPApiClient" that handles that and can be constructed from a Region, but not assuming there is a singular such that associated with a Region.

File: src/sentry/types/region.py [hunk 6]
```diff
@@ -1,8 +1,120 @@
...
+        self._by_id = {r.id: r for r in self.regions}
+
+    @classmethod
+    def load_from_config(cls) -> RegionMapping:
+        from django.conf import settings
+
+        # For now, assume that all region configs can be taken in through Django
+        # settings. We may investigate other ways of delivering those configs in
+        # production.
+        return cls(settings.SENTRY_REGION_CONFIG)
```
Comment (corps) [inline src/sentry/types/region.py:68]: 👍

File: src/sentry/types/region.py [hunk 7]
```diff
@@ -1,8 +1,120 @@
...
+
+    def get_by_name(self, name: str) -> Region | None:
+        """Look up a region by name."""
+        return self._by_name.get(name)
```
Comment (corps) [inline src/sentry/types/region.py:72]: Is there possibly some merit to explicitly throwing exception with useful error message when lookups fail?  The problem is almost certainly consumers will all either be written to do the same -- check for null and throw a useful error message, or worse, do no checking and result in unuseful NoneType exceptions.

File: src/sentry/types/region.py (hunk 3 above)
Comment (corps) [reply src/sentry/types/region.py:116]: What about an api like,

```
get_local_region() -> str
get_all_regions() -> List[str]
get_region(key: str) -> Region
```

? I don't think we need the Mapping object directly.

Comment (corps) [reply src/sentry/types/region.py:116]: IE: Essentially flatten the mapping object into module level functions and don't expose it directly.

File: tests/sentry/utils/test_snowflake.py [hunk 8]
```diff
@@ -57,3 +63,27 @@ def test_out_of_region_sequences(self):
             generate_snowflake_id("test_redis_key")
 
         assert str(context.value) == "No available ID"
+
+    @freeze_time(CURRENT_TIME)
+    @mock.patch("sentry.utils.snowflake.get_region_mapping")
```
Comment (corps) [inline tests/sentry/utils/test_snowflake.py:68]: How about using `with override_settings` as the test target?  Helps test deeper the load from settings mechanism as well.  Minor nit though.

**Comment by corps** [comment]
Thank you for getting this started, honestly I can imagine using this framework for features in the near future.

File: src/sentry/types/region.py (hunk 3 above)
Comment (markstory) [reply src/sentry/types/region.py:116]: Having a public API made of functions feels better than a global API of shared global data. I recognize that the functions are just wrapping the global state. But the functions *could* be cloning data so that the global state can't be mutated.

File: src/sentry/types/region.py (hunk 7 above)
Comment (markstory) [reply src/sentry/types/region.py:72]: Exceptions is better imo. A region being missing and an organization not belonging to a region when the application is expecting a region feel like 'exceptional' scenarios.

Comment (RyanSkonnord) [reply src/sentry/types/region.py]: 👍 I was on the fence myself. Switched to exceptions in 03d8911f4335954756269c4e28e3cd74e711707b

File: src/sentry/types/region.py [hunk 9]
```diff
@@ -1,8 +1,120 @@
...
+from typing import TYPE_CHECKING, Iterable
+
+from sentry.silo import SiloMode
+
+if TYPE_CHECKING:
+    from sentry.models import Organization
+
+
+class RegionCategory(Enum):
+    GEOGRAPHIC = "GEOGRAPHIC"
```
[the remark this answers is not in the document]
Comment (RyanSkonnord) [reply src/sentry/types/region.py]: Changed (33f63667de66578f0e014556a69a0d3dec5056a2)

File: tests/sentry/types/test_region.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
+from django.test import override_settings
+
+from sentry.silo import SiloMode
+from sentry.testutils import TestCase
+from sentry.types.region import Region, RegionCategory, RegionMapping
+
+
+class RegionMappingTest(TestCase):
```
Comment (AniketDas-Tekky) [inline tests/sentry/types/test_region.py:19]: Could you provide an example test of mocking the region?

**Review by AniketDas-Tekky — APPROVED** [review]

File: src/sentry/types/region.py (hunk 3 above)
Comment (RyanSkonnord) [reply src/sentry/types/region.py]: Module-level interfaces make my Java brain sad but I see how it's nicer to abstract away the mapping object. 😅

Part of the reason I didn't like that originally, was that `override_settings` wouldn't play nice with having the regions cached. (I don't believe the caching is premature optimization, because the mapping could otherwise produce O(n<sup>2</sup>) costs once we have a growing number of single-tenant regions.) In addition to injecting the region values, you would have to mess around with mocks in the `sentry.types.region` module, which usually would be two steps removed from whatever you're trying to test, [as is the case in `test_snowflake`](https://github.com/getsentry/sentry/blob/33f63667de66578f0e014556a69a0d3dec5056a2/tests/sentry/utils/test_snowflake.py#L68).

My solution is [a helper named `override_region` in `testutils`](https://github.com/getsentry/sentry/blob/ceafdf59d6daa5d663ed2d1949f1a340f4e5be2c/src/sentry/testutils/region.py), which hides away the module-level mocking for reuse in other tests. I think [the current `test_snowflake` implementation](https://github.com/getsentry/sentry/blob/ceafdf59d6daa5d663ed2d1949f1a340f4e5be2c/tests/sentry/utils/test_snowflake.py#L69-L73) works out pretty nicely, where it only has to inject `SENTRY_REGION` but doesn't have to directly mock any functions.

File: tests/sentry/utils/test_snowflake.py (hunk 8 above)
Comment (RyanSkonnord) [reply tests/sentry/utils/test_snowflake.py]: As I mentioned [here](https://github.com/getsentry/sentry/pull/39253#discussion_r980508451), `override_settings` isn't enough by itself, but I think the current implementation is as good.

File: src/sentry/types/region.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (RyanSkonnord) [reply src/sentry/types/region.py]: If it's a monolith, there will be no region object at all.

File: src/sentry/utils/snowflake.py (hunk 4 above)
Comment (RyanSkonnord) [reply src/sentry/utils/snowflake.py]: As discussed elsewhere, I prefer not to represent a "default region" when what we really mean is "there are no regions in my entire system (which may be a self-hosted monolith)".

> I think we should avoid SnowflakeIDs if we haven't set the region config just yet.

The plan is to have snowflake IDs be used for all organizations and projects, for the sake of parity, regardless of whether the system is multi-region. So it would be normal for future self-hosted systems to be generating snowflake IDs with zeroes where the region IDs would be if they had regions. (CC @maxiuyuan)

In contrast to not wanting a region *object* as a "default value", the snowflake IDs force us to represent both the "region" and "no-region" cases with numbers. Therefore, I agree with reserving 0 for "unregioned" and documenting it in code (and error-checking it in `Region.__init__`) . We'd then count actual regions from 1.

File: src/sentry/types/region.py (hunk 3 above)
Comment (corps) [reply src/sentry/types/region.py]: Is it also possible to "clear" this cached value via a pytest after hook that can be installed globally?

File: tests/sentry/types/test_region.py (hunk 10 above)
Comment (RyanSkonnord) [reply tests/sentry/types/test_region.py:19]: Are you looking for anything more than what's in ceafdf59d6daa5d663ed2d1949f1a340f4e5be2c?

File: src/sentry/utils/snowflake.py (hunk 4 above)
Comment (RyanSkonnord) [reply src/sentry/utils/snowflake.py]: Went ahead and did this in b52156c66c75117cab3403d0ce7b860a4d716783. Let me know if anyone feels strongly about the North American region being 0.

File: src/sentry/types/region.py (hunk 3 above)
Comment (RyanSkonnord) [reply src/sentry/types/region.py]: Not without using a more sophisticated cache than `functools.lru_cache`, as far as I know.

**Review by corps — APPROVED** [review]
